snap/gadget.go

```diff
@@ -69,6 +72,85 @@ type VolumeContent struct {
 	Unpack bool `yaml:"unpack"`
 }
 
+// GadgetConnect describes an interface connection requested by the gadget
+// between seeded snaps. The syntax is of a mapping like:
+//
+//  plug: (<plug-snap-id>|system):plug
+//  [slot: (<slot-snap-id>|system):slot]
+//
+// "system" indicates a system plug or slot.
+// Fully omitting the slot part indicates a system slot with the same name
+// as the plug.
+type GadgetConnection struct {
+	Plug GadgetConnectionPlug `yaml:"plug"`
+	Slot GadgetConnectionSlot `yaml:"slot"`
+}
+
+type GadgetConnectionPlug struct {
+	SnapID string
+	Plug   string
+}
+
+func (gcplug *GadgetConnectionPlug) Empty() bool {
+	return gcplug.SnapID == "" && gcplug.Plug == ""
+}
+
+func (gcplug *GadgetConnectionPlug) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var s string
+	if err := unmarshal(&s); err != nil {
+		return err
+	}
+	snapID, name, err := parseSnapIDColonName(s)
+	if err != nil {
+		return fmt.Errorf("in gadget connection plug: %v", err)
+	}
+	gcplug.SnapID = snapID
+	gcplug.Plug = name
+	return nil
+}
+
+type GadgetConnectionSlot struct {
+	SnapID string
+	Slot   string
+}
+
+func (gcslot *GadgetConnectionSlot) Empty() bool {
+	return gcslot.SnapID == "" && gcslot.Slot == ""
+}
+
+func (gcslot *GadgetConnectionSlot) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var s string
+	if err := unmarshal(&s); err != nil {
+		return err
+	}
+	snapID, name, err := parseSnapIDColonName(s)
+	if err != nil {
+		return fmt.Errorf("in gadget connection slot: %v", err)
+	}
+	gcslot.SnapID = snapID
+	gcslot.Slot = name
+	return nil
+}
+
+func parseSnapIDColonName(s string) (snapID, name string, err error) {
+	parts := strings.Split(s, ":")
+	if len(parts) == 2 {
+		snapID = parts[0]
+		name = parts[1]
+	}
+	if snapID == "" || name == "" {
+		return "", "", fmt.Errorf(`expected "(<snap-id>|system):name" not %q`, s)
+	}
+	return snapID, name, nil
+}
+
+func systemOrSnapID(s string) bool {
+	if s != "system" && len(s) != 32 {
```

Comment:
Where does 32 come from? Presumably it's the length agreed with the store? I couldn't find any references to this number in the existing code, perhaps it's worth having constant for it somewhere (snap/info.go?).